File: BOJ/python3/1935.py
'''
후위 표기식2
https://www.acmicpc.net/problem/1935
'''
import sys
import time
start = time.time()  # 시작 시간
sys.stdin = open("input.txt", "r")  # input.txt 파일로 입력


N = int(input())
postfix = input()
nums = [None for _ in range(N)]
for i in range(N):
    nums[i] = int(input())

operator = ["+", "-", "*", "/"]
operand = [None, None]
stack = []
for ch in postfix:
    if ch not in operator:
        # 연산자가 아닐경우 스택에 저장
        stack.append(ch)
    else:
        # 연산자면 두 수를 꺼내와 계산 후 다시 저장
        for i in range(2):
            n = stack.pop()
            if isinstance(n, str):
                n = nums[ord(n) - 65]  # 알파벳 -> 숫자

            operand[-1+i] = n

        a, b = operand[0], operand[1]
        if ch == "+":
            c = a + b
        elif ch == "-":
            c = a - b
        elif ch == "*":
            c = a * b
        elif ch == "/":
            c = a / b

        stack.append(c)

print(f"{stack[0]:.2f}")

# 현재시각 - 시작시간 = 실행 시간(sec)
print("--------------------")
print(f"Elapsed Time: {round(time.time() - start, 3)}s")

File: BOJ/nodejs/1935.js
/*
후위 표기식2
boj.kr/1935
 */
const start = new Date().getTime();

const fs = require("fs");
const input = fs.readFileSync("input.txt").toString().split("\n");
//const input = fs.readFileSync("/dev/stdin").toString().split("\n");

const N = 1 * input[0];
const postfix = input[1].trimEnd();
const nums = new Array(26).fill(null);
for (let n = 0; n < N; n++) {
  nums[n] = 1 * input[2 + n];
}

const stack = new Array();
for (let ch of postfix) {
  if (ch.charCodeAt() >= 65 && ch.charCodeAt() <= 90) {
    stack.push(nums[ch.charCodeAt() - 65]);
  } else {
    let rt = stack.pop();
    let lt = stack.pop();
    switch (ch) {
      case "+":
        stack.push(lt + rt);
        break;
      case "-":
        stack.push(lt - rt);
        break;
      case "*":
        stack.push(lt * rt);
        break;
      case "/":
        stack.push(lt / rt);
        break;
    }
  }
}

let ans = stack.pop().toFixed(2);
console.log(ans);

const end = new Date().getTime();
console.log("--------------------");
console.log(`Elapsed Time: ${(new Date().getTime() - start) / 1000}s`);
